File: src-tauri/src/types/domain.rs
/// Domain/business entity types — projects, sessions, metrics, pagination.

use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::messages::ParsedMessage;

// =============================================================================
// Project & Session
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Project {
    pub id: String,
    pub path: String,
    pub name: String,
    pub sessions: Vec<String>,
    pub created_at: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub most_recent_session: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PhaseTokenBreakdown {
    pub phase_number: u32,
    pub contribution: u64,
    pub peak_tokens: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub post_compaction: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Session {
    pub id: String,
    pub project_id: String,
    pub project_path: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub todo_data: Option<Value>,
    pub created_at: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub first_message: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message_timestamp: Option<String>,
    pub has_subagents: bool,
    pub message_count: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub is_ongoing: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub git_branch: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata_level: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context_consumption: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub compaction_count: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub phase_breakdown: Option<Vec<PhaseTokenBreakdown>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub custom_title: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub agent_name: Option<String>,
}

// =============================================================================
// Metrics
// =============================================================================

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionMetrics {
    pub duration_ms: f64,
    pub total_tokens: u64,
    pub input_tokens: u64,
    pub output_tokens: u64,
    pub cache_read_tokens: u64,
    pub cache_creation_tokens: u64,
    pub message_count: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cost_usd: Option<f64>,
    /// Primary model used in this session (most frequent assistant model)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
}

// =============================================================================
// Parsed Session (result of full parse)
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParsedSession {
    pub messages: Vec<ParsedMessage>,
    pub metrics: SessionMetrics,
    pub task_calls: Vec<super::messages::ToolCall>,
    pub by_type: MessagesByType,
    pub sidechain_messages: Vec<ParsedMessage>,
    pub main_messages: Vec<ParsedMessage>,
    pub custom_title: Option<String>,
    pub agent_name: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MessagesByType {
    pub user: Vec<ParsedMessage>,
    pub real_user: Vec<ParsedMessage>,
    pub internal_user: Vec<ParsedMessage>,
    pub assistant: Vec<ParsedMessage>,
    pub system: Vec<ParsedMessage>,
    pub other: Vec<ParsedMessage>,
}

// =============================================================================
// Pagination
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PaginatedSessionsResult {
    pub sessions: Vec<Session>,
    pub next_cursor: Option<String>,
    pub has_more: bool,
    pub total_count: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionsPaginationOptions {
    #[serde(default = "default_true")]
    pub include_total_count: bool,
    #[serde(default = "default_true")]
    pub prefilter_all: bool,
    #[serde(default = "default_deep")]
    pub metadata_level: String,
}

fn default_true() -> bool {
    true
}

fn default_deep() -> String {
    "deep".to_string()
}

impl Default for SessionsPaginationOptions {
    fn default() -> Self {
        Self {
            include_total_count: true,
            prefilter_all: true,
            metadata_level: "deep".to_string(),
        }
    }
}

File: internal/domain/domain.go
// Package domain holds the shared DTOs ported from src-tauri/src/types/.
// It is a pure leaf: it imports only the stdlib (encoding/json), never any
// other internal package, so the parsing/analysis packages can depend on it
// without forming an import cycle.
//
// JSON encoding mirrors serde exactly (the parity gate diffs byte-for-byte):
//   - Rust `rename_all = "camelCase"` → explicit camelCase `json:` tags.
//   - `Option<T>` + `skip_serializing_if="Option::is_none"` → `*T` + `,omitempty`
//     (omits the key when nil, matching serde).
//   - `Option<T>` WITHOUT skip → `*T` and NO `,omitempty` (serde emits `null`).
//   - frontend-iterated slices are constructed as `[]T{}`, never left nil.
package domain

// Project — a project folder under ~/.claude/projects.
type Project struct {
	ID                string   `json:"id"`
	Path              string   `json:"path"`
	Name              string   `json:"name"`
	Sessions          []string `json:"sessions"`
	CreatedAt         float64  `json:"createdAt"`
	MostRecentSession *float64 `json:"mostRecentSession,omitempty"`
}

// PhaseTokenBreakdown — per-compaction-phase token contribution.
type PhaseTokenBreakdown struct {
	PhaseNumber    uint32  `json:"phaseNumber"`
	Contribution   uint64  `json:"contribution"`
	PeakTokens     uint64  `json:"peakTokens"`
	PostCompaction *uint64 `json:"postCompaction,omitempty"`
}

// Session — session-level metadata.
type Session struct {
	ID               string                 `json:"id"`
	ProjectID        string                 `json:"projectId"`
	ProjectPath      string                 `json:"projectPath"`
	TodoData         RawValue               `json:"todoData,omitempty"`
	CreatedAt        float64                `json:"createdAt"`
	FirstMessage     *string                `json:"firstMessage,omitempty"`
	MessageTimestamp *string                `json:"messageTimestamp,omitempty"`
	HasSubagents     bool                   `json:"hasSubagents"`
	MessageCount     uint32                 `json:"messageCount"`
	IsOngoing        *bool                  `json:"isOngoing,omitempty"`
	GitBranch        *string                `json:"gitBranch,omitempty"`
	MetadataLevel    *string                `json:"metadataLevel,omitempty"`
	ContextConsumption *uint64              `json:"contextConsumption,omitempty"`
	CompactionCount  *uint32                `json:"compactionCount,omitempty"`
	PhaseBreakdown   *[]PhaseTokenBreakdown `json:"phaseBreakdown,omitempty"`
	CustomTitle      *string                `json:"customTitle,omitempty"`
	AgentName        *string                `json:"agentName,omitempty"`
}

// SessionMetrics — token/cost/duration aggregation. Default-derivable.
type SessionMetrics struct {
	DurationMs          float64 `json:"durationMs"`
	TotalTokens         uint64  `json:"totalTokens"`
	InputTokens         uint64  `json:"inputTokens"`
	OutputTokens        uint64  `json:"outputTokens"`
	CacheReadTokens     uint64  `json:"cacheReadTokens"`
	CacheCreationTokens uint64  `json:"cacheCreationTokens"`
	MessageCount        uint32  `json:"messageCount"`
	CostUsd             *float64 `json:"costUsd,omitempty"`
	Model               *string  `json:"model,omitempty"`
}

// ParsedSession — result of a full parse (process_messages output).
type ParsedSession struct {
	Messages         []ParsedMessage `json:"messages"`
	Metrics          SessionMetrics  `json:"metrics"`
	TaskCalls        []ToolCall      `json:"taskCalls"`
	ByType           MessagesByType  `json:"byType"`
	SidechainMessages []ParsedMessage `json:"sidechainMessages"`
	MainMessages     []ParsedMessage `json:"mainMessages"`
	// custom_title / agent_name have NO skip_serializing_if → emit null when absent.
	CustomTitle *string `json:"customTitle"`
	AgentName   *string `json:"agentName"`
}

// MessagesByType — messages bucketed by role/category.
type MessagesByType struct {
	User         []ParsedMessage `json:"user"`
	RealUser     []ParsedMessage `json:"realUser"`
	InternalUser []ParsedMessage `json:"internalUser"`
	Assistant    []ParsedMessage `json:"assistant"`
	System       []ParsedMessage `json:"system"`
	Other        []ParsedMessage `json:"other"`
}

// PaginatedSessionsResult — a page of sessions.
type PaginatedSessionsResult struct {
	Sessions   []Session `json:"sessions"`
	NextCursor *string   `json:"nextCursor"` // no skip → null when absent
	HasMore    bool      `json:"hasMore"`
	TotalCount uint32    `json:"totalCount"`
}
